stl/inc/xstring

```diff
@@ -2432,6 +2440,106 @@ public:
+        : _Mypair(_One_then_variadic_args_t(), _Left._Getal()) {
+        auto& _My_data    = _Mypair._Myval2;
+        auto& _Left_data  = _Left._Mypair._Myval2;
+        auto& _Right_data = _Right._Mypair._Myval2;
+        _Left_data._Orphan_all();
+        _Right_data._Orphan_all();
+        const auto _Left_size  = _Left_data._Mysize;
+        const auto _Right_size = _Right_data._Mysize;
+
+        const auto _Left_capacity  = _Left_data._Myres;
+        const auto _Right_capacity = _Right_data._Myres;
+        // overflow is OK due to max_size() checks:
+        const auto _New_size = static_cast<size_type>(_Left_size + _Right_size);
+        if (_Right_size <= _Left_capacity - _Left_size && _Left_capacity >= _Right_capacity) {
+            // take _Left's buffer, max_size() is OK because the result fits in _Left's capacity
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Left, bool_constant<_Can_memcpy_val>{});
+            const auto _Ptr = _My_data._Myptr();
+            _Traits::copy(_Ptr + _My_data._Mysize, _Right_data._Myptr(), _Right_size);
+            _My_data._Mysize = _New_size;
+            _Traits::assign(_Ptr[_New_size], _Elem());
+            return;
+        }
+
+        if (_Allocators_equal(_Getal(), _Right._Getal()) && _Left_size <= _Right_capacity - _Right_size) {
+            // take _Right's buffer, max_size() is OK because the result fits in _Right's capacity
+            // At this point, we have tested:
+            // !(_Right_size <= _Left_capacity - _Left_size && _Left_capacity >= _Right_capacity)
+            //     && _Left_size <= _Right_capacity - _Right_size
+            // therefore: (demorgan)
+            // (_Right_size > _Left_capacity - _Left_size || _Left_capacity < _Right_capacity)
+            //     && _Left_size <= _Right_capacity - _Right_size
+            // therefore: (distribute)
+            // (_Right_size > _Left_capacity - _Left_size && _Left_size <= _Right_capacity - _Right_size)
+            // || (_Left_capacity < _Right_capacity && _Left_size <= _Right_capacity - _Right_size)
+            // therefore: either the result fits in _Right's capacity but not in _Left's capacity, so _Right must have
+            //     had more capacity, OR we explicitly tested that _Right had more capacity
+            // therefore: _Right must be in large mode (since it's capacity is larger than the minimum), so save that
+            //     branch
+            _STL_INTERNAL_CHECK(_Right_data._Large_string_engaged());
+            _My_data._Alloc_proxy(_GET_PROXY_ALLOCATOR(_Alty, _Getal())); // throws, hereafter nothrow in this block
+            _Take_contents(_Right, bool_constant<_Can_memcpy_val>{});
+            const auto _Ptr = _Unfancy(_My_data._Bx._Ptr);
+            _Traits::move(_Ptr + _Left_size, _Ptr, _My_data._Mysize);

```

Comment:
Same comment about avoiding a null terminator assignment at the end - you could simply move down the freshly taken null terminator here. This will also improve locality a little, hopefully.

Reply:
Shouldn't improve locality at all; after all we just wrote the character before that null terminator so that cache location is hot.

Reply:
You were performing the following actions: take right's buffer, move down right's data (without the null terminator), copy in left's data to the beginning of the buffer, then go back to the end of the buffer and write the null terminator. If you had scribbled the null terminator immediately after moving down right's data, then I would agree that there is no locality change.

Reply:
Ah, I see, because this is the move around case. OK.